### PR TITLE
Add Scala support to AndroidLibraryDescription.

### PR DESCRIPTION
**Pull request by zbsz**
Added AndroidLibraryCompiler abstraction to allow for different jvm languages be used in android library.
Added language parameter to AndroidLibrary arguments.

It should be quite easy to later add Kotlin and Groovy support using the same abstraction.

### CODE REVIEW

**Comment by kageiit** [comment]
@dsyang has some plans for android kotlin support in #857

Adding them to ensure this is compatible

**Comment by dsyang** [comment]
I haven't implemented (or really thought closely) about doing this for
kotlin yet. This sounds like a much better approach. If it works for scala
that's awesome and we should merge it. I'm sure we can do something similar
for kotlin.

Thanks for doing this @zbsz!!

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      AndroidLibraryDescription.Arg arg,
+      JavacOptions javacOptions,
+      BuildRuleResolver resolver) {
+
+    return new JavacToJarStepFactory(javacOptions, new BootClasspathAppender());
+  }
+}
+
+
+class ScalaAndroidLibraryCompiler extends AndroidLibraryCompiler {
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Why leave these here instead of in their own file (or package)?

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            .addAll(scalaBuckConfig.getCompilerFlags())
+            .addAll(arg.extraArguments.get())
+            .addAll(Iterables.transform(
+                resolver.getAllRules(scalaBuckConfig.getCompilerPlugins()),
+                new Function<BuildRule, String>() {
+                  @Override
+                  public String apply(BuildRule input) {
+                    return "-Xplugin:" + input.getPathToOutput();
+                  }
+                }))
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Maybe we should extract this to a utility/static method instead of duplicating between here and `ScalaLibraryDescription`

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    new Function<String, Path>() {
+                      @Override
+                      public Path apply(String input) {
+                        return new File(input).toPath();
+                      }
+                    }
+                );
+          }
+        }
+    );
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: nit: feels like this can (and should) be consolidated

File: src/com/facebook/buck/android/AndroidLibraryDescription.java [hunk 4]
```diff
@@ -164,17 +187,42 @@ public Arg createUnpopulatedConstructorArg() {
...
     public Optional<SourcePath> manifest;
     public Optional<String> resourceUnionPackage;
     public Optional<String> finalRName;
+
+    @Hint(isInput = false)
+    public Optional<JvmLanguage> language;
+  }
+
+  public enum JvmLanguage {
+    JAVA, SCALA
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: nit: split into 2 lines

File: src/com/facebook/buck/android/AndroidLibraryDescription.java [hunk 5]
```diff
@@ -47,17 +52,20 @@
...
   public static final BuildRuleType TYPE = BuildRuleType.of("android_library");
 
   private static final Flavor DUMMY_R_DOT_JAVA_FLAVOR =
       AndroidLibraryGraphEnhancer.DUMMY_R_DOT_JAVA_FLAVOR;
 
   private final JavacOptions defaultOptions;
+  private final ScalaBuckConfig scalaBuckConfig;
 
-  public AndroidLibraryDescription(JavacOptions defaultOptions) {
+  public AndroidLibraryDescription(JavacOptions defaultOptions, ScalaBuckConfig scalaConfig) {
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: I wonder if it's possible to do this without hardcoding `ScalaBuckConfig` into the default `AndroidLibraryDescription`. It's not going to be scalable when we start implementing support for kotlin (and other jvm languages).

**Comment by zbsz** [comment]
@dsyang thanks for comments. Do you have suggestions how to solve AndroidPlatformTarget and ScalaBuckConfig issues? I'm not very familiar with the code and didn't see any obvious way to do it better, so any tips would be appreciated.

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 3 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Shoulf I add 'AndroidPlatformTarget' to 'BuildContext'? Or is there some better way?

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 1 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Moved to separate files.

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 2 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Fixed

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 3 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Replaced android boot classpath supplier in `BuildContext` with `AndroidPlatformTarget` supplier. Seems better this way.

File: src/com/facebook/buck/android/AndroidLibraryDescription.java (hunk 5 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: Not sure how to improve that, we could introduce some additional class for this, something like `JvmLanguagesConfig`, but it's just moving the same things from arguments list to fields. Any better suggestions?

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 2 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Actually, I could extract this whole method, will do that later.

File: src/com/facebook/buck/android/AndroidLibraryDescription.java (hunk 5 above)
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: What if we had a `JvmLanguagesConfig` interface or base class that the`Scala/Kotlin/BuckConfig` classes implement.  Then we use `JvmLanguagesConfig` here. 

2) Pass in an android compiler instead of a buck config. Your `getCompiler()` implementation doesn't look like it uses anything inside this class and can probably be pulled out into some selector class that choose

Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: 1) `JvmLanguagesConfig` would need to include almost everything that is in language specific configs, I mean there doesn't seem to be a common subset of settings that would be enough to instantiate any `AndroidLibraryCompiler`
2) `getCompiler` uses `language` param, so it can not be created before `AndroidLibraryDescription` 

3)
Other option would be to have class providing suppliers for all configs, something like:
```
abstract class AndroidLibraryConfig {
    Supplier<ScalaBuckConfig> getScalaConfigSuplier();
    Supplier<KotlinBuckConfig> getKotlinConfigSupplier();
    ...
}
```
All methods would have default implementations throwing exceptions, so there is no need to specify all configs every time.
What do you think about that?

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 2 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: To achieve better reuse it could be good to replace `AndroidLibraryCompiler` with common `JvmLibraryCompiler`, this would be more general compiler provider without references to android (placed in `jvm.common`), then `ScalaLIbraryDescription` (and others) would use it the same way `AndroidLibraryDescription` does.
I'm just afraid that this will end up being a bigger refactoring, and I'm not sure if I want to do it as a part of this PR.

File: src/com/facebook/buck/android/AndroidLibraryDescription.java (hunk 5 above)
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: @Coneko @mikekap @grumpyjames what do you guys think?

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 2 above)
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Yea I agree. definitely sounds like something we should to in a separate PR

File: src/com/facebook/buck/android/AndroidLibraryDescription.java (hunk 5 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: Just tried adding `AndroidLibraryCompilerFactory`, in last commit. Not sure if that's worth it, it ads some indirection, library description doesn't need to know all the configs, but factory will still need them.

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import com.facebook.buck.rules.BuildRuleResolver;
+import com.facebook.buck.rules.CellPathResolver;
+import com.facebook.buck.rules.ImplicitDepsInferringDescription;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+
+import java.nio.file.Path;
+
+
+@SuppressWarnings("unused")
```
Comment (Coneko) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Why is this necessary?

Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Ant complains about unused parameters in default implementation, for example: 'The value of the parameter resolver is not used'.

File: src/com/facebook/buck/rules/KnownBuildRuleTypes.java [hunk 7]
```diff
@@ -525,7 +526,7 @@ static Builder createBuilder(
 
     KotlinBuckConfig kotlinBuckConfig = new KotlinBuckConfig(config);
 
-    ScalaBuckConfig scalaConfig = new ScalaBuckConfig(config);
+    final ScalaBuckConfig scalaConfig = new ScalaBuckConfig(config);
```
Comment (Coneko) [inline src/com/facebook/buck/rules/KnownBuildRuleTypes.java]: This doesn't need to be final, I assume it's a leftover from some earlier version of the code.

File: test/com/facebook/buck/android/AndroidLibraryIntegrationTest.java [hunk 8]
```diff
@@ -51,4 +51,20 @@ public void testAndroidLibraryDoesNotUseTransitiveResources() throws IOException
     result.assertFailure();
     assertTrue(result.getStderr().contains("package R does not exist"));
   }
+
+  @Test
```
Comment (Coneko) [inline test/com/facebook/buck/android/AndroidLibraryIntegrationTest.java]: All the other scala integration tests have increased timeouts because of how long they take on CI. Adjust these accordingly as well.

File: src/com/facebook/buck/rules/BuildContext.java [hunk 9]
```diff
@@ -78,8 +50,8 @@ public String get() {
   public abstract BuckEventBus getEventBus();
 
   @Value.Default
-  public Supplier<String> getAndroidBootclasspathSupplier() {
```
Comment (Coneko) [inline src/com/facebook/buck/rules/BuildContext.java]: Please extract this change to a separate PR.

Comment (zbsz) [inline src/com/facebook/buck/rules/BuildContext.java]: Created #915

File: src/com/facebook/buck/rules/KnownBuildRuleTypes.java (hunk 7 above)
Comment (zbsz) [inline src/com/facebook/buck/rules/KnownBuildRuleTypes.java]: fixed, yup forgot to revert that

File: test/com/facebook/buck/android/testdata/android_project/scala/com/sample/lib/BUCK.fixture [hunk 10]
```diff
@@ -0,0 +1,35 @@
+android_library(
+  name = 'lib',
+  srcs = glob(['*.scala']),
+  language = 'SCALA',
```
Comment (dsyang) [inline test/com/facebook/buck/android/testdata/android_project/scala/com/sample/lib/BUCK.fixture]: can we also add a test that says compilation will fail if the language parameter isn't something recognized? ("java", "scala",etc...)

File: test/com/facebook/buck/rules/BuildContextTest.java [hunk 11]
```diff
@@ -17,85 +17,29 @@
...
 
-import java.io.File;
-import java.nio.file.Path;
-import java.nio.file.Paths;
-import java.util.List;
-
 public class BuildContextTest {
 
-  @Test
-  public void testGetAndroidBootclasspathSupplierWithAndroidPlatformTarget() {
```
Comment (dsyang) [inline test/com/facebook/buck/rules/BuildContextTest.java]: I realize this was extracted into #915 but do we have an equivalent test that covers this case? Seems weird to just remove test cases instead of moving them to a different unit test. I could be wrong but that PR didn't seem to have one.

File: src/com/facebook/buck/android/AndroidLibraryDescription.java [hunk 12]
```diff
@@ -172,10 +198,28 @@ public boolean hasFlavors(ImmutableSet<Flavor> flavors) {
...
+        .findDepsForTargetFromConstructorArgs(buildTarget, cellRoots, constructorArg);
+  }
+
   @SuppressFieldNotInitialized
   public static class Arg extends JavaLibraryDescription.Arg {
     public Optional<SourcePath> manifest;
     public Optional<String> resourceUnionPackage;
     public Optional<String> finalRName;
+
+    @Hint(isInput = false)
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: (Just trying to understand the codebase better) why did we need to add this?

File: src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.android;
+
+public interface AndroidLibraryCompilerFactory {
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java]: Javadoc

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+public abstract class AndroidLibraryCompiler
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: Javadoc

File: test/com/facebook/buck/android/testdata/android_project/scala/com/sample/lib/BUCK.fixture (hunk 10 above)
Comment (zbsz) [inline test/com/facebook/buck/android/testdata/android_project/scala/com/sample/lib/BUCK.fixture]: Added in `AndroidLibraryIntegrationTest`, although I'm not sure if that's exactly right, this fails on Arg parsing level, so it's not reported as build failure, but `runBuckBuild` throws an exception.

File: test/com/facebook/buck/rules/BuildContextTest.java (hunk 11 above)
Comment (zbsz) [inline test/com/facebook/buck/rules/BuildContextTest.java]: You're right, I was just lazy, added test in `AndroidLibraryDescriptionTest`

File: src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java (hunk 13 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java]: Added small description, any improvements will be welcome.

File: test/com/facebook/buck/jvm/java/DefaultJavaLibraryTest.java [hunk 15]
```diff
@@ -1382,7 +1384,9 @@ private BuildRule createJavaLibraryRule(ProjectFilesystem projectFilesystem)
           /* providedDeps */ ImmutableSortedSet.<BuildRule>of(),
           /* abiJar */ new FakeSourcePath("abi.jar"),
           /* additionalClasspathEntries */ ImmutableSet.<Path>of(),
-          options.build(),
+          options,
+          options.trackClassUsage(),
+          new JavacToJarStepFactory(options, new BootClasspathAppender()),
```
Comment (dsyang) [inline test/com/facebook/buck/jvm/java/DefaultJavaLibraryTest.java]: @shs96c Correct me if I'm wrong but your TODO from line 1352 should be relatively simple to fix right? We're making c/f/b/android/BootClasspathAppender public because of this test and I don't think we need to.

@zbsz if you think it's easy to update go for it. Otherwise lets create an issue.

File: src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.android;
+
+import com.facebook.buck.android.AndroidLibraryDescription.JvmLanguage;
+
+/**
+ * Factory providing implementations of {@link AndroidLibraryCompiler}
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java]: How about this?
```
/**
  * Factory providing implementations of {@link AndroidLibraryCompiler}
  * for the specified {@code language}.
  * {@link AndroidLibraryDescription} uses this factory to handle multiple JVM languages.
  * Implementations should provide a compiler implementation for every {@link JvmLanguage}. 
  * See {@link DefaultAndroidLibraryCompiler}
  */
```

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+import com.facebook.buck.rules.BuildRuleResolver;
+import com.facebook.buck.rules.CellPathResolver;
+import com.facebook.buck.rules.ImplicitDepsInferringDescription;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+
+import java.nio.file.Path;
+
+/**
+ * Jvm compiler abstraction for android library.
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]:
```
/**
  * Jvm compiler abstraction for android.
  * Implementations of this class are used in {@link AndroidLibraryDescription} to provide
 * the actual compilation step.
  * This allows us to use different compilers for different {@link JvmLanguage}.
  */

File: src/com/facebook/buck/android/AndroidLibraryDescription.java [hunk 18]
```diff
@@ -172,10 +197,26 @@ public boolean hasFlavors(ImmutableSet<Flavor> flavors) {
...
+
   @SuppressFieldNotInitialized
   public static class Arg extends JavaLibraryDescription.Arg {
     public Optional<SourcePath> manifest;
     public Optional<String> resourceUnionPackage;
     public Optional<String> finalRName;
+    public Optional<JvmLanguage> language;
+  }
+
+  public enum JvmLanguage {
```
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: fixed

File: src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java (hunk 16 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompilerFactory.java]: replaced

File: src/com/facebook/buck/android/AndroidLibraryCompiler.java (hunk 17 above)
Comment (zbsz) [inline src/com/facebook/buck/android/AndroidLibraryCompiler.java]: replaced

**Comment by zbsz** [comment]
Rebased with master.
Android integration test keeps failing on Travis, seems to be cause by memory limits, not sure how to fix that.
